cli/src/commands/fix.rs

```diff
@@ -48,106 +69,252 @@ pub(crate) fn cmd_fix(
 ) -> Result<(), CommandError> {
     let mut workspace_command = command.workspace_helper(ui)?;
     let root_commits: Vec<Commit> = workspace_command
-        .parse_union_revsets(&args.sources)?
+        .parse_union_revsets(if args.sources.is_empty() {
+            &[RevisionArg::AT]
+        } else {
+            &args.sources
+        })?
         .evaluate_to_commits()?
         .try_collect()?;
     workspace_command.check_rewritable(root_commits.iter().ids())?;
 
-    let mut tx = workspace_command.start_transaction();
+    let mut tx: WorkspaceCommandTransaction = workspace_command.start_transaction();
 
-    // Collect all FileIds we're going to format and which commits they appear in
+    // Collect all of the unique tool inputs we're going to use. Tools should be
+    // deterministic, and should not consider outside information, so it is safe to
+    // deduplicate inputs that correspond to multiple files or commits. This is
+    // typically more efficient, but it does prevent certain use cases like
+    // providing commit IDs as inputs to be inserted into files. We also need to
+    // record the mapping between tool inputs and paths/commits, to efficiently
+    // rewrite the commits later.
+    //
+    // If a path is being fixed in a particular commit, it must also be fixed in all
+    // that commit's descendants. We do this as a way of propagating changes,
+    // under the assumption that it is more useful than performing a rebase and
+    // risking merge conflicts. In the case of code formatters, rebasing wouldn't
+    // reliably produce well formatted code anyway. Deduplicating inputs helps
+    // to prevent quadratic growth in the number of tool executions required for
+    // doing this in long chains of commits with disjoint sets of modified files.
     let commits: Vec<_> = RevsetExpression::commits(root_commits.iter().ids().cloned().collect())
         .descendants()
         .evaluate_programmatic(tx.base_repo().as_ref())?
         .iter()
         .commits(tx.repo().store())
         .try_collect()?;
-    let mut file_ids = HashSet::new();
-    let mut commit_paths: HashMap<CommitId, Vec<RepoPathBuf>> = HashMap::new();
+    let mut all_tool_inputs: HashSet<ToolInput> = HashSet::new();
+    let mut commit_tool_inputs: HashMap<CommitId, HashMap<RepoPathBuf, Vec<ToolInput>>> =
+        HashMap::new();
```

Comment:
nit: Not sure we should change it, but it's a bit ironic that `all_tool_inputs` has fewer tool inputs than `commit_tool_inputs` :)